libbeat/publisher/pipeline/output_test.go

```diff
@@ -0,0 +1,181 @@
+// Licensed to Elasticsearch B.V. under one or more contributor
+// license agreements. See the NOTICE file distributed with
+// this work for additional information regarding copyright
+// ownership. Elasticsearch B.V. licenses this file to you under
+// the Apache License, Version 2.0 (the "License"); you may
+// not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+package pipeline
+
+import (
+	"math/rand"
+	"sync"
+	"testing"
+	"time"
+
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+
+	"github.com/elastic/beats/v7/libbeat/common/atomic"
+	"github.com/elastic/beats/v7/libbeat/logp"
+	"github.com/elastic/beats/v7/libbeat/outputs"
+	"github.com/elastic/beats/v7/libbeat/publisher"
+	"github.com/elastic/beats/v7/libbeat/publisher/queue"
+)
+
+func TestPublish(t *testing.T) {
+	tests := map[string]struct {
+		client outputs.Client
+	}{
+		"client": {
+			&mockClient{},
+		},
+		"network_client": {
+			&mockNetworkClient{},
+		},
+	}
+
+	for name, test := range tests {
+		t.Run(name, func(t *testing.T) {
+			rand.Seed(time.Now().UnixNano())
+
+			wqu := makeWorkQueue()
+			makeClientWorker(nilObserver, wqu, test.client)
+
+			numEvents := atomic.MakeInt(0)
+			var wg sync.WaitGroup
+			for batchIdx := 0; batchIdx <= randIntBetween(25, 200); batchIdx++ {
+				wg.Add(1)
+				go func() {
+					defer wg.Done()
+					batch := randomBatch(50, 150, wqu)
+
+					numEvents.Add(len(batch.Events()))
+
+					wqu <- batch
+				}()
+			}
+			wg.Wait()
+
+			// Give some time for events to be published
+			time.Sleep(time.Duration(numEvents.Load()*3) * time.Microsecond)
+
+			// Make sure that all events have eventually been published
+			c := test.client.(interface{ Published() int })
+			assert.Equal(t, numEvents.Load(), c.Published())
+		})
+	}
+}
+
+func TestPublishWithClose(t *testing.T) {
+	rand.Seed(time.Now().UnixNano())
+
+	wqu := makeWorkQueue()
+	client := &mockNetworkClient{}
```

Comment:
The reason I'm only using a `outputs.NetworkClient` here is because that will result in a `netClientWorker` below (instead of a `clientWorker`). And it appears that only in the case of a `netClientWorker` do we cancel the batch if the worker was closed before the batch was published: https://github.com/elastic/beats/blob/8d431691401478bc8d6ed3c80e64b6092e2a332a/libbeat/publisher/pipeline/output.go#L120.

Are we missing a similar check for `clientWorker` inside this loop? https://github.com/elastic/beats/blob/8d431691401478bc8d6ed3c80e64b6092e2a332a/libbeat/publisher/pipeline/output.go#L69-L75

Reply:
For reloading, I would say we're missing the Cancel for client worker. `clientworker` is only used for console and file output, which I don't think are available via fleet, but better to have the code correct.
Could the reason be that the console and file output do not support a 'Stop/Close' method? Maybe we need to touch those as well?

Reply:
Just noticed the difference is the `Connect` method. Yeah, clientWorker should also cancel batches in case it is replaced.